elmslie-core/src/main/java/vivid/money/elmslie/core/store/ElmStore.kt

```diff
@@ -17,123 +16,72 @@ class ElmStore<Event : Any, State : Any, Effect : Any, Command : Any>(
-        return Disposable {
-            effectListeners -= onEffectEmission
-            effectBuffer.clear()
-            if (isStarted && effectListeners.isEmpty()) stopBuffering()
-        }
-    }
-
-    private fun events(onEventTriggering: (Event) -> Unit): Disposable {
-        eventListeners += onEventTriggering
-        return Disposable { eventListeners -= onEventTriggering }
+        _isStarted.set(false)
+        storeScope.cancel()
     }
 
-    override fun <ChildEvent : Any, ChildState : Any, ChildEffect : Any> addChildStore(
-        childStore: Store<ChildEvent, ChildEffect, ChildState>,
-        eventMapper: (parentEvent: Event) -> ChildEvent?,
-        effectMapper: (parentState: State, childEffect: ChildEffect) -> Effect?,
-        stateReducer: (parentState: State, childState: ChildState) -> State
-    ): Store<Event, Effect, State> {
-        disposables.addAll(
-            // We won't lose any state or effects since they're cached
-            { childStore.stop() },
-            events { eventMapper(it)?.let(childStore::accept) },
-            childStore.effects { effectMapper(currentState, it)?.let(::dispatchEffect) },
-            childStore.states { dispatchState(stateReducer(currentState, it)) },
-        )
-        childStore.start()
-        return this
-    }
-
-    private fun dispatchState(state: State) = synchronized(stateLock) {
-        stateInternal.set(state)
-        stateListeners.forEach { it(state) }
-    }
+    override fun states(): Flow<State> = statesFlow.asStateFlow()
 
-    private fun dispatchEffect(effect: Effect) {
-        logger.debug("New effect: $effect")
-        effectListeners.forEach { it(effect) }
-    }
+    override fun effects(): Flow<Effect> = effectsFlow.asSharedFlow()
 
     private fun dispatchEvent(event: Event) {
-        executor.submit {
+        storeScope.launch {
             try {
                 logger.debug("New event: $event")
-                eventListeners.forEach { it(event) }
-                val result = reducer.reduce(event, currentState)
-                dispatchState(result.state)
-                result.effects.forEach(::dispatchEffect)
-                result.commands.forEach(::executeCommand)
+                val (state, effects, commands) = reducer.reduce(event, currentState)
+                statesFlow.value = state
+                effects.forEach(::dispatchEffect)
+                commands.forEach(::executeCommand)
             } catch (t: Throwable) {
                 logger.fatal("You must handle all errors inside reducer", t)
             }
         }
     }
 
-    private fun executeCommand(command: Command) = try {
-        logger.debug("Executing command: $command")
-        disposables += actor.execute(command, ::dispatchEvent, { logger.nonfatal(error = it) })
-    } catch (t: Throwable) {
-        logger.fatal("Unexpected actor error", t)
+    private fun dispatchEffect(effect: Effect) {
+        storeScope.launch {
```

Comment:
For [executeCommand] I suggest to leave as is, because each command returns a flow and we should create for each flow separate coroutine to ensure that all commands would be handled.